[AGW] mobilityD: Add support for static IPs

pshelar opened the pull request:
## Summary

    This patch check subscriberDB for statically assigned
    IP address for given subscriber before allocating IP
    address. It uses GRPC channel to communicate to local
    SubscriberDB service.

## Test Plan

`make test`
`make integ_test`

pshelar committed:
mobilityD: Add support for static IPs

This patch check subscriberDB for statically assigned
IP address for given subscrivber before allocating IP
address. It uses GRPC channel to communicate to local
SubscriberDB service.

Signed-off-by: Pravin B Shelar <pbshelar@fb.com>

pshelar commented:
The test failure is due to a mock issue, please ignore it for now.

xjtian commented on lte/gateway/python/magma/mobilityd/serialize_utils.py:
```diff
@@ -51,6 +51,26 @@ def _desc_state_proto_to_str(proto):
     return state
 
 
+def _desc_type_str_to_proto(type):
```
you should either rebase these changes out of this PR in favor of #2155 or abandon that PR and just land those changes with this PR

pshelar replied:
That is first patch, so if you review by commits in this pull request you can ignore it.

xjtian commented on lte/gateway/python/magma/mobilityd/ip_address_man.py:
```diff
@@ -432,6 +436,19 @@ def _try_set_recycle_timer(self):
+            return None
+        return IPDesc(ip=ip_addr, state=IPState.ALLOCATED,
+                      sid=sid, ip_block=ip_addr, ip_type=IPType.STATIC)
+
```
this logic belongs to the IPAllocator, not this class. 2 different options here:

1. If static IP enabled means static IP *only*, then create a new implementation of IPAllocator for static IP configuration.
2. If static IP enabled means allocate a static IP when it's configured and another method if it's not, then create a new IPAllocator implementation which can wrap a "fallback" IP Allocator.

High level point here is that it's not this component's responsibility to know about the details of IP allocation, static or otherwise

pshelar replied:
done, Added wrapper class.

xjtian commented on lte/gateway/python/magma/mobilityd/subscriberdb_client.py:
```diff
@@ -0,0 +1,77 @@
+SERVICE_NAME = "subscriberdb"
+
+
+def get_subscriber_ip(sid: str) -> ip_address:
```
you should encapsulate this inside a class that you inject into your new IPAllocator implementation. This will make the unit test setup a lot more straightforward because you can just control the return value of `get_subscriber_ip` instead of mocking out the subscriberdb stub at a global level

pshelar replied:
done.

pshelar committed:
rename ip_allocator_static to ip_allocator_pool

Signed-off-by: Pravin B Shelar <pbshelar@fb.com>

ardzoht commented on lte/gateway/python/magma/mobilityd/ip_allocator_static.py:
```diff
@@ -18,190 +18,75 @@
-                 ip_state_map: IpDescriptorMap,
-                 sid_ips_map: defaultdict):
-        """ Initializes a new IP allocator
+    def __init__(self, rpc_stub, ip_allocator: IPAllocator):
```
rpc_stub sounds generic, can we update it to subscriberdb_rpc_stub? also it would be great updating the paremeter with typing (SubscriberDBStub)

ardzoht commented on lte/gateway/python/magma/mobilityd/rpc_servicer.py:
```diff
@@ -52,11 +52,12 @@ def _get_ip_block(ip_block_str):
     """ gRPC based server for the IPAllocator. """
 
-    def __init__(self, mconfig, config):
+    def __init__(self, mconfig, config, rpc_stub=None):
```
same here, let's be explicit on the service this rpc stub is referring to

ardzoht commented on lte/gateway/python/magma/mobilityd/subscriberdb_client.py:
```diff
@@ -0,0 +1,64 @@
+            apn_name = ''
+
+            if '.' in sid:
+                apn_name = sid.split('.')[1]
```
`imsi, apn_name = sid.split('.')` to unpack the values from the splitted list since we know the expected num of values returned

ardzoht commented on lte/gateway/python/magma/mobilityd/ip_address_man.py:
```diff
@@ -153,19 +159,25 @@ def __init__(self,
+                                           dhcp_store=dhcp_store,
+                                           gw_info=self._dhcp_gw_info)
+
+        if self.static_ip_enabled:
```
I feel this is a little messy since we are instantiating an ip_allocator already given the type of it, and then we're injecting it as a dependency to another class that seems to wrap the logic in it. 

Can we include this wrapping extension into the already defined classes? Or maybe rename the IpAllocatorStatic to something like IPAllocatorStaticWrapper since it looks more like an extension of the ip allocator with subscriberdb client rather than another type of IP allocator.

pshelar replied:
In V1 extended IPAddressManager to check for static IP, so I do not want to extend existing IP allocator classes, given that would duplicate the logic. I can rename the class.

xjtian reviewed (approved):
I landed the other PR with the proto changes so you'll have to rebase

pshelar commented:
> I landed the other PR with the proto changes so you'll have to rebase

Rebased on latest master. Only difference in the patch is that I have added mobilityD service dependency.
